Arm/Arm.cs

```diff
@@ -0,0 +1,290 @@
+
+        public void MoveTo(float X, float Y, float Z)
+        {
+            const float LearningRate = 0.0001f;
+            const int Iterations = 40;
+            for (int j = 0; j < Params.Length; j++)
+            {
+                if (Params[j].Fixed)
+                {
+                    var (CurX, CurY, CurZ) = ForwardKinematics();
+                    float PrevSum = j > 0 ? PitchSums[j - 1] : 0;
+                    CurrentPitches[j] = EnsureStandardForm((Params[j].FixedAngle - PrevSum));
+                }
+                else
+                {
+                    for (int i = 0; i < Iterations; i++)
+                    {
+                        var (CurX, CurY, CurZ) = ForwardKinematics();
+                        //float Cost = (CurX - X) * (CurX - X) + (CurY - Y) * (CurY - Y) + (CurZ - Z) * (CurZ - Z);
+                        var (GradRoll, GradPitch, GradYaw) = CalcGrad(X, Y, Z, CurX, CurY, CurZ, j);
+
+                        GradRoll *= LearningRate;
+                        GradPitch *= LearningRate;
+                        GradYaw *= LearningRate;
+
+                        TempRolls[j] = CurrentRolls[j] - GradRoll;
+                        TempPitches[j] = CurrentPitches[j] - GradPitch;
+                        TempYaws[j] = CurrentYaws[j] - GradYaw;
+
+                        /*
+                        Console.WriteLine("GradRoll:  " + GradRoll);
+                        Console.WriteLine("GradPitch: " + GradPitch);
+                        Console.WriteLine("GradYaw:   " + GradYaw);
+                        */
+
+                        float[] tmp = CurrentRolls;
+                        CurrentRolls = TempRolls;
+                        TempRolls = tmp;
+
+                        tmp = CurrentPitches;
+                        CurrentPitches = TempPitches;
+                        TempPitches = tmp;
+
+                        tmp = CurrentYaws;
+                        CurrentYaws = TempYaws;
+                        TempYaws = tmp;
+                    }
+                }
+                CurrentRolls[j] = Math.Max(Math.Min(CurrentRolls[j], Params[j].MaxRoll), Params[j].MinRoll);
+                CurrentPitches[j] = Math.Max(Math.Min(CurrentPitches[j], Params[j].MaxPitch), Params[j].MinPitch);
+                CurrentYaws[j] = Math.Max(Math.Min(CurrentYaws[j], Params[j].MaxYaw), Params[j].MinYaw);
+
+                for (int k = 0; k < CurrentRolls.Length; k++)
+                    Console.WriteLine($"({CurrentRolls[k]}, {CurrentPitches[k]}, {CurrentYaws[k]})");
+                Console.WriteLine();
+            }
+        }
+
+        public Arm(params ArmPart[] Parameters)
+        {
+            this.Params = Parameters;
+            this.CurrentRolls = new float[Parameters.Length];
+            this.CurrentPitches = new float[Parameters.Length];
+            this.CurrentYaws = new float[Parameters.Length];
+
+            Random r = new Random();
+            for (int i = 0; i < Params.Length; i++)
```

Comment:
Randomization is needed because of everything is 0, gradients will be 0 and nothing will be updated. This is a common technique in machine learning, particularly neural networks.